libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c

```diff
@@ -564,15 +621,36 @@ TickType_t xWriteTimeOut_ms = ipconfigSOCK_DEFAULT_SEND_BLOCK_TIME;
 
 				ulIPAddress = 0UL;
 
-				if( FreeRTOS_sendto( xDNSSocket, pucUDPPayloadBuffer, xPayloadLength, FREERTOS_ZERO_COPY, &xAddress, sizeof( xAddress ) ) != 0 )
+				if( FreeRTOS_sendto( xDNSSocket, pucUDPPayloadBuffer, uxPayloadLength, FREERTOS_ZERO_COPY, &xAddress, sizeof( xAddress ) ) != 0 )
 				{
 					/* Wait for the reply. */
 					lBytes = FreeRTOS_recvfrom( xDNSSocket, &pucUDPPayloadBuffer, 0, FREERTOS_ZERO_COPY, &xAddress, &ulAddressLength );
 
 					if( lBytes > 0 )
 					{
+					BaseType_t xExpected;
+					DNSMessage_t *pxDNSMessageHeader = ( DNSMessage_t * ) pucUDPPayloadBuffer;
+
+						/* See if the identifiers match. */
+						if( uxIdentifier == ( TickType_t ) pxDNSMessageHeader->usIdentifier )
+						{
+							xExpected = pdTRUE;
+						}
+						else
+						{
+							/* The reply was not expected. */
+							xExpected = pdFALSE;
+						}
+
 						/* The reply was received.  Process it. */
-						ulIPAddress = prvParseDNSReply( pucUDPPayloadBuffer, lBytes, xIdentifier );
+					#if( ipconfigDNS_USE_CALLBACKS == 0 )
+						/* It is useless to analyse the unexpected reply
```

Comment:
> Is uxIdentifier provided only when ipconfigDNS_USE_CALLBACKS is true?

`uxIdentifier` is always provided, it is non-zero and "random".

> If uxIdentifier is always provided, is there any reason not to check for a match?

Every incoming packet will be checked for a match:

    if( uxIdentifier == ( TickType_t ) pxDNSMessageHeader->usIdentifier )

At this point, a reply has been received.
If the identifier in the reply matches with the request that was just sent, `xExpected` will set `true`.
An expected reply **must** be analysed by `prvParseDNSReply`.

What if a reply is received that was not expected?

- With callback: analyse it because it might be a reply to an outstanding request
- No callback: skip this reply

I summarised this in this phrase:

    It is useless to analyse the unexpected reply
    unless asynchronous look-ups are enabled

Please feel free to modify the text to make it clearer.